CesiumGeometry/src/IntersectionTests.cpp

```diff
@@ -31,31 +32,97 @@ IntersectionTests::rayPlane(const Ray& ray, const Plane& plane) noexcept {
   return ray.getOrigin() + ray.getDirection() * t;
 }
 
-bool IntersectionTests::pointInTriangle2D(
+bool IntersectionTests::pointInTriangle(
     const glm::dvec2& point,
     const glm::dvec2& triangleVertA,
     const glm::dvec2& triangleVertB,
     const glm::dvec2& triangleVertC) noexcept {
-
+  // Define the triangle edges.
   const glm::dvec2 ab = triangleVertB - triangleVertA;
-  const glm::dvec2 ab_perp(-ab.y, ab.x);
   const glm::dvec2 bc = triangleVertC - triangleVertB;
-  const glm::dvec2 bc_perp(-bc.y, bc.x);
   const glm::dvec2 ca = triangleVertA - triangleVertC;
+
+  // Get the vector perpendicular to each triangle edge in the same 2D plane).
+  const glm::dvec2 ab_perp(-ab.y, ab.x);
+  const glm::dvec2 bc_perp(-bc.y, bc.x);
   const glm::dvec2 ca_perp(-ca.y, ca.x);
 
+  // Get vectors from triangle corners to point.
   const glm::dvec2 av = point - triangleVertA;
   const glm::dvec2 cv = point - triangleVertC;
 
+  // Project the point vectors onto the perpendicular vectors.
   const double v_proj_ab_perp = glm::dot(av, ab_perp);
   const double v_proj_bc_perp = glm::dot(cv, bc_perp);
   const double v_proj_ca_perp = glm::dot(cv, ca_perp);
 
-  // This will determine in or out, irrespective of winding.
+  // The projections determine in what direction the point is from the triangle
+  // edge. This will determine in or out, irrespective of winding.
   return (v_proj_ab_perp >= 0.0 && v_proj_ca_perp >= 0.0 &&
           v_proj_bc_perp >= 0.0) ||
          (v_proj_ab_perp <= 0.0 && v_proj_ca_perp <= 0.0 &&
           v_proj_bc_perp <= 0.0);
 }
 
+bool IntersectionTests::pointInTriangle(
+    const glm::dvec3& point,
+    const glm::dvec3& triangleVertA,
+    const glm::dvec3& triangleVertB,
+    const glm::dvec3& triangleVertC) noexcept {
+  glm::dvec3 unused;
+  return IntersectionTests::pointInTriangle(
+      point,
+      triangleVertA,
+      triangleVertB,
+      triangleVertC,
+      unused);
+}
+
+bool IntersectionTests::pointInTriangle(
+    const glm::dvec3& point,
+    const glm::dvec3& triangleVertA,
+    const glm::dvec3& triangleVertB,
+    const glm::dvec3& triangleVertC,
+    glm::dvec3& barycentricCoordinates) noexcept {
+  // PERFORMANCE: If optimization is necessary in the future, there are
+  // algorithms that avoid length computations (which involve square root
+  // operations).
+
+  // Define the triangle edges.
+  const glm::dvec3 ab = triangleVertB - triangleVertA;
+  const glm::dvec3 bc = triangleVertC - triangleVertB;
+
+  const glm::dvec3 triangleNormal = glm::cross(ab, bc);
+  const double lengthSquared = glm::length2(triangleNormal);
+  if (lengthSquared < CesiumUtility::Math::Epsilon8) {
+    // Don't handle invalid triangles.
+    return false;
+  }
+
+  // Technically this is the area of the parallelogram between the vectors,
+  // but triangles are 1/2 of the area, and the division will cancel out later.
+  const double triangleAreaInv = 1.0 / glm::sqrt(lengthSquared);
```

Comment:
You don't need to calculate the total area of the triangle. The barycentric coordinates must sum to 1.0, so you can add up your computed area coordinates (lengths of the cross products below) and then normalize by that at the end.

Reply:
Sorry I'm not sure I follow. How would you normalize the sum of the areas without computing the total area of the triangle?

I'm using the first two area sizes to deduce the third one with subtraction. If I only have the first two, then I still would have to deduce the first one by doing `totalArea - firstTriangle - secondTriangle`, and then normalizing it.

Reply:
I saw your other comment and agreed with it, and I added a check for `triangleCAPRatio` being negative. Can I just clarify how you are suggesting I rewrite the code? This is what I think I'm deducing from your comment:

```
 // Technically this is the area of the parallelogram between the vectors,
  // but triangles are 1/2 of the area, and the division will cancel out later.
  const glm::dvec3 ap = point - triangleVertA;
  const double areaABP =
      glm::length(glm::cross(ab, ap));

  const glm::dvec3 bp = point - triangleVertB;
  const double areaBCP =
      glm::length(glm::cross(bc, bp));

  const glm::dvec3 cp = point - triangleVertC;
  const double areaCAP =
      glm::length(glm::cross(ca, cp));
      
  const double denominator = areaABP + areaBCP + areaCAP;
  const double triangleABPRatio = areaABP / denominator;
  // etc.
  
  if (triangleABPRatio < 0 || triangleBCPRatio < 0 || triangleCAPRatio < 0)
    return false;
```

Reply:
Sorry just realizing my error with the above (at the bottom), the areas will all be positive so they're never going to be negative. So I feel like I still have to check for the total area because, normalized, the sum could be greater than 1?

Reply:
Ack! Of course the areas aren't negative, so feel free to ignore most of my comments. I was fixated on the 2D case where the result of the cross product can be negative.